### PR TITLE
Allow layer-wise recompute

### PR DESCRIPTION
**Pull request by pengwa**
### Allow layer-wise recompute 

Early, we need users/developers to specify the subgraphs to recompute, now we introduced a more user-friendly way to enable recompute for all detected stashed activation recomputation subgraphs. This scarifies getting the best configs while makes it easier to support user requirements when they switches from PyTorch per-layer gradient checkpoint to ORTModule. 

`ORTMODULE_MEMORY_OPT_LEVEL` is introduced to control the usage, by default, it is 0, e.g. `USER_SPECIFIED`, all subgraphs definedin `ORTMODULE_MEMORY_OPT_CONFIG` will be recomputed. So this is compatible to existing recompute usage in ORTModule integrated models. 

Using `ORTMODULE_MEMORY_OPT_LEVEL=1`, we will enable all recompute plans detected, so those configs in `ORTMODULE_MEMORY_OPT_CONFIG` will not be respected any more. 

Add Unit Tests using 3 layer blooms. 

https://github.com/microsoft/onnxruntime/blob/pengwa/add_aggresive_recompute/docs/Memory_Optimizer.md

### CODE REVIEW

**Comment by AdamLouly** [comment]
- How does ORT finds all the recompute opportunities in this implementation?
- Are there numbers that shows how much memory optimization we have against how much perf we lost?

**Comment by pengwa** [comment]
> * How does ORT finds all the recompute opportunities in this implementation?
It is here: https://github.com/microsoft/onnxruntime/blob/f3369a8bf87190552ad551a6de56df01cccf7a62/orttraining/orttraining/core/optimizer/memory_optimizer/memory_insight.cc#L249C23-L249C53. 

Find those stashed activations that are used by backward operators. Put all those activations as candidates; For each candidate, https://github.com/microsoft/onnxruntime/blob/f3369a8bf87190552ad551a6de56df01cccf7a62/orttraining/orttraining/core/optimizer/memory_optimizer/memory_insight.cc#L272C9-L272C30 will check whether it is recomputable, and how the subgraph looks like,. 

> * Are there numbers that shows how much memory optimization we have against how much perf we lost?

No, that's the long term goal to have all those feature ready, to help dynamically choose a good plan for users.

File: onnxruntime/core/graph/graph_viewer.cc [hunk 1]
```diff
@@ -35,6 +35,17 @@ struct PriorityNodeCompare {
       return n1->Priority() > n2->Priority();
     }
 
+    // nodes of forward pass will be output first
+    auto n1_attrs = n1->GetAttributes();
+    auto n2_attrs = n2->GetAttributes();
+    int64_t n1_is_forward = static_cast<int64_t>(n1_attrs.find(kBackwardNodeAttributeName) == n1_attrs.cend()) ||
```
Comment (askhade) [inline onnxruntime/core/graph/graph_viewer.cc:41]: I dont quite understand how this value is being calculated. What will be n1_is_forward when "(n1_attrs.find(kBackwardNodeAttributeName) == n1_attrs.cend())" is True and when it is False.

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py [hunk 2]
```diff
@@ -451,9 +449,11 @@ def _create_execution_agent(self):
             )
 
             self._runtime_inspector.memory_ob.find_memory_optimization_opportunity(
-                execution_agent, self._runtime_options.memory_optimizer_config, self._runtime_options.probe_level
+                execution_agent, self._runtime_options
             )
 
+            self._logger.debug("Completed find_memory_optimization_opportunity.")
```
Comment (askhade) [inline orttraining/orttraining/python/training/ortmodule/_training_manager.py:455]: Is this necessary? Can this be moved inside find_memory_optimization_opportunity?

File: onnxruntime/core/graph/graph_viewer.cc (hunk 1 above)
Comment (pengwa) [reply onnxruntime/core/graph/graph_viewer.cc:41]: You can check line 25 and 26. If the function return true, n2 will be output first, e.g. before n1 in the topo order. So in this case, if n2 is forward pass, but n1 is not, then n2 should be before n1.

File: orttraining/orttraining/test/optimizer/memory_optimizer_test.cc [hunk 3]
```diff
@@ -135,13 +137,184 @@ TEST(MemoryOptimizerTests, TileRecompute) {
...
+    if (!node_ptr) continue;  // Node was removed.
+
+    if (std::find(bw_nodes_in_expected_order.begin(), bw_nodes_in_expected_order.end(), node_ptr) !=
+        bw_nodes_in_expected_order.end()) {
+      nodes_in_topological_order.push_back(j);
+      j++;
+    }
+  }
+
+  for (size_t i = 1; i < nodes_in_topological_order.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (pengwa) [reply orttraining/orttraining/test/optimizer/memory_optimizer_test.cc:310]: added in python tests.

**Review by zhijxu-MS — APPROVED** [review]

**Review by askhade — APPROVED** [review]

**Comment by pengwa** [comment]
Thank you @askhade, @zhijxu-MS !!
